server/controllers/finance/journal/core.js

```diff
@@ -1,198 +1,156 @@
+    .addQuery(
+      `SET @transId = (SELECT CONCAT(abbr, IFNULL(MAX(increment), 1)) AS id FROM (
+        SELECT project.abbr, MAX(FLOOR(SUBSTR(trans_id, 4))) + 1 AS increment
+        FROM posting_journal JOIN project ON posting_journal.project_id = project.id
+        WHERE posting_journal.project_id = @projectId
+      UNION
+        SELECT project.abbr, MAX(FLOOR(SUBSTR(trans_id, 4))) + 1 AS increment
+        FROM general_ledger JOIN project ON general_ledger.project_id = project.id
+        WHERE general_ledger.project_id = @projectId)c
+      );`
+    )
+
+    // set up the @enterpriseCurrencyId
+    .addQuery(`
+      SET @enterpriseCurrencyId = (
+        SELECT currency_id FROM enterprise WHERE id = @enterpriseId
+      );
+    `)
+
+    // set up the @rate SQL variable
+    .addQuery(`
+      SET @rate = (
+        SELECT rate FROM exchange_rate
+        WHERE enterprise_id = @enterpriseId
+          AND currency_id = @currencyId
+          AND date <= @date
+      ORDER BY date DESC
+      LIMIT 1);
+    `)
+
+    // if the currency is the enterprise currency, we will set @exchange to 1,
+    // otherwise it is 1/@rate
+    .addQuery(
+      `SET @exchange = (SELECT IF(@currencyId = @enterpriseCurrencyId, 1, 1/@rate));`
+    )
+
+    // error handling query - uses stored procedure PostingJournalErrorHandler
+    // to make sure we have all the SQL variables properly set (not NULL);
+    .addQuery(`
+      CALL PostingJournalErrorHandler(
+        @enterpriseId, @projectId, @fiscalId, @periodId, @exchange, @date
+      );
+    `);
+
+  return transaction;
 };
 
-// get the exchange rate for a given date
-queries.myExchangeRate = function (date) {
-  'use strict';
-
-  // expects a mysql-compatible date
-  var sql =
-    'SELECT exchange_rate.enterprise_id, exchange_rate.currency_id, exchange_rate.rate, ' +
-    'enterprise.currency_id AS \'enterprise_currency_id\' ' +
-    'FROM exchange_rate JOIN currency ON ' +
-      'exchange_rate.currency_id = currency.id ' +
-      'JOIN enterprise ON enterprise.id = exchange_rate.enterprise_id ' +
-    'WHERE exchange_rate.date = DATE(?);';
-
-  return db.exec(sql, [date])
-  .then(function (rows) {
-    if (rows.length === 0) {
-      throw new Error('No exchange rate found for date : ' + date);
-    }
-
-    var store = new Store();
-    rows.forEach(function (line) {
-      store.post({ id : line.currency_id, rate : line.rate });
-      store.post({ id : line.enterprise_currency_id, rate : 1});
-    });
-
-    return q(store);
-  });
+/**
+ * Core Error Handler
+ *
+ * This function will catch database errors generated during the posting
+ * transaction, and properly convert them into JavaScript errors to be
+ * sent to the client.
+ *
+ * @param {Error} error - the error object thrown by MySQL.
+ * @returns {Promise} handled - a rejected promise with converted error.
+ */
+exports.handler = function handler(error) {
+  let handled;
+
+  switch (error.sqlState) {
```

Comment:
Is there a reason this uses a `switch` and not a map of codes like [this](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/server/config/codes.js)? This is not a problem I'm just wondering if there was a reason for the different model. Should the other be updated? 

My first thought was that this could be handled in a middleware like other database errors however having a core handler may come in useful for considering further edge cases in the future. No conclusion on this one!